File: test/src/CartPoleUtil.cpp
// Copyright (c) Sleipnir contributors

#include "CartPoleUtil.hpp"

// https://underactuated.mit.edu/acrobot.html#cart_pole
//
// θ is CCW+ measured from negative y-axis.
//
// q = [x, θ]ᵀ
// q̇ = [ẋ, θ̇]ᵀ
// u = f_x
//
// M(q)q̈ + C(q, q̇)q̇ = τ_g(q) + Bu
// M(q)q̈ = τ_g(q) − C(q, q̇)q̇ + Bu
// q̈ = M⁻¹(q)(τ_g(q) − C(q, q̇)q̇ + Bu)
//
//        [ m_c + m_p  m_p l cosθ]
// M(q) = [m_p l cosθ    m_p l²  ]
//
//           [0  −m_p lθ̇ sinθ]
// C(q, q̇) = [0       0      ]
//
//          [     0      ]
// τ_g(q) = [-m_p gl sinθ]
//
//     [1]
// B = [0]

inline constexpr double m_c = 5.0;  // Cart mass (kg)
inline constexpr double m_p = 0.5;  // Pole mass (kg)
inline constexpr double l = 0.5;    // Pole length (m)
inline constexpr double g = 9.806;  // Acceleration due to gravity (m/s²)

Eigen::Vector<double, 4> CartPoleDynamicsDouble(
    const Eigen::Vector<double, 4>& x, const Eigen::Vector<double, 1>& u) {
  auto q = x.segment(0, 2);
  auto qdot = x.segment(2, 2);
  auto theta = q(1);
  auto thetadot = qdot(1);

  //        [ m_c + m_p  m_p l cosθ]
  // M(q) = [m_p l cosθ    m_p l²  ]
  Eigen::Matrix<double, 2, 2> M{
      {m_c + m_p, m_p * l * cos(theta)},              // NOLINT
      {m_p * l * cos(theta), m_p * std::pow(l, 2)}};  // NOLINT

  double detM = M(0, 0) * M(1, 1) - M(0, 1) * M(1, 0);
  Eigen::Matrix<double, 2, 2> Minv{{M(1, 1) / detM, -M(0, 1) / detM},
                                   {-M(1, 0) / detM, M(0, 0) / detM}};

  //           [0  −m_p lθ̇ sinθ]
  // C(q, q̇) = [0       0      ]
  Eigen::Matrix<double, 2, 2> C{
      {0, -m_p * l * thetadot * sin(theta)},  // NOLINT
      {0, 0}};

  //          [     0      ]
  // τ_g(q) = [-m_p gl sinθ]
  Eigen::Vector<double, 2> tau_g{{0}, {-m_p * g * l * sin(theta)}};  // NOLINT

  //     [1]
  // B = [0]
  constexpr Eigen::Matrix<double, 2, 1> B{{1}, {0}};

  // q̈ = M⁻¹(q)(τ_g(q) − C(q, q̇)q̇ + Bu)
  Eigen::Vector<double, 4> qddot;
  qddot.segment(0, 2) = qdot;
  qddot.segment(2, 2) = Minv * (tau_g - C * qdot + B * u);
  return qddot;
}

sleipnir::VariableMatrix CartPoleDynamics(const sleipnir::VariableMatrix& x,
                                          const sleipnir::VariableMatrix& u) {
  auto q = x.Segment(0, 2);
  auto qdot = x.Segment(2, 2);
  auto theta = q(1);
  auto thetadot = qdot(1);

  //        [ m_c + m_p  m_p l cosθ]
  // M(q) = [m_p l cosθ    m_p l²  ]
  sleipnir::VariableMatrix M{
      {m_c + m_p, m_p * l * cos(theta)},              // NOLINT
      {m_p * l * cos(theta), m_p * std::pow(l, 2)}};  // NOLINT

  auto detM = M(0, 0) * M(1, 1) - M(0, 1) * M(1, 0);
  sleipnir::VariableMatrix Minv{{M(1, 1) / detM, -M(0, 1) / detM},
                                {-M(1, 0) / detM, M(0, 0) / detM}};

  //           [0  −m_p lθ̇ sinθ]
  // C(q, q̇) = [0       0      ]
  sleipnir::VariableMatrix C{{0, -m_p * l * thetadot * sin(theta)},  // NOLINT
                             {0, 0}};

  //          [     0      ]
  // τ_g(q) = [-m_p gl sinθ]
  sleipnir::VariableMatrix tau_g{{0}, {-m_p * g * l * sin(theta)}};  // NOLINT

  //     [1]
  // B = [0]
  constexpr Eigen::Matrix<double, 2, 1> B{{1}, {0}};

  // q̈ = M⁻¹(q)(τ_g(q) − C(q, q̇)q̇ + Bu)
  sleipnir::VariableMatrix qddot{4};
  qddot.Segment(0, 2) = qdot;
  qddot.Segment(2, 2) = Minv * (tau_g - C * qdot + B * u);
  return qddot;
}

File: jormungandr/test/cart_pole_util.py
import math

import jormungandr.autodiff as autodiff
from jormungandr.autodiff import VariableMatrix, Variable
import numpy as np

# https://underactuated.mit.edu/acrobot.html#cart_pole
#
# θ is CCW+ measured from negative y-axis.
#
# q = [x, θ]ᵀ
# q̇ = [ẋ, θ̇]ᵀ
# u = f_x
#
# M(q)q̈ + C(q, q̇)q̇ = τ_g(q) + Bu
# M(q)q̈ = τ_g(q) − C(q, q̇)q̇ + Bu
# q̈ = M⁻¹(q)(τ_g(q) − C(q, q̇)q̇ + Bu)
#
#        [ m_c + m_p  m_p l cosθ]
# M(q) = [m_p l cosθ    m_p l²  ]
#
#           [0  −m_p lθ̇ sinθ]
# C(q, q̇) = [0       0      ]
#
#          [     0      ]
# τ_g(q) = [-m_p gl sinθ]
#
#     [1]
# B = [0]

m_c = 5.0  # Cart mass (kg)
m_p = 0.5  # Pole mass (kg)
l = 0.5  # Pole length (m)
g = 9.806  # Acceleration due to gravity (m/s²)


def cart_pole_dynamics_double(x, u):
    q = x[:2, :]
    qdot = x[2:, :]
    theta = q[1, 0]
    thetadot = qdot[1, 0]

    #        [ m_c + m_p  m_p l cosθ]
    # M(q) = [m_p l cosθ    m_p l²  ]
    M = np.array(
        [
            [m_c + m_p, m_p * l * math.cos(theta)],
            [m_p * l * math.cos(theta), m_p * l**2],
        ]
    )

    #           [0  −m_p lθ̇ sinθ]
    # C(q, q̇) = [0       0      ]
    C = np.array([[0.0, -m_p * l * thetadot * math.sin(theta)], [0.0, 0.0]])

    #          [     0      ]
    # τ_g(q) = [-m_p gl sinθ]
    tau_g = np.array([[0.0], [-m_p * g * l * math.sin(theta)]])

    #     [1]
    # B = [0]
    B = np.array([[1], [0]])

    # q̈ = M⁻¹(q)(τ_g(q) − C(q, q̇)q̇ + Bu)
    qddot = np.empty((4, 1))
    qddot[:2, :] = qdot
    qddot[2:, :] = np.linalg.solve(M, tau_g - C @ qdot + B @ u)
    return qddot


def cart_pole_dynamics(x, u):
    q = x[:2, :]
    qdot = x[2:, :]
    theta = q[1, 0]
    thetadot = qdot[1, 0]

    #        [ m_c + m_p  m_p l cosθ]
    # M(q) = [m_p l cosθ    m_p l²  ]
    M = VariableMatrix(
        [
            [Variable(m_c + m_p), m_p * l * autodiff.cos(theta)],
            [m_p * l * autodiff.cos(theta), Variable(m_p * l**2)],
        ]
    )

    #           [0  −m_p lθ̇ sinθ]
    # C(q, q̇) = [0       0      ]
    C = VariableMatrix(
        [
            [Variable(0.0), -m_p * l * thetadot * autodiff.sin(theta)],
            [Variable(0.0), Variable(0.0)],
        ]
    )

    #          [     0      ]
    # τ_g(q) = [-m_p gl sinθ]
    tau_g = VariableMatrix([[Variable(0.0)], [-m_p * g * l * autodiff.sin(theta)]])

    #     [1]
    # B = [0]
    B = np.array([[1], [0]])

    # q̈ = M⁻¹(q)(τ_g(q) − C(q, q̇)q̇ + Bu)
    qddot = VariableMatrix(4, 1)
    qddot[:2, :] = qdot
    qddot[2:, :] = autodiff.solve(M, tau_g - C @ qdot + B @ u)
    return qddot
